src/connector/src/sink/pulsar.rs

```diff
@@ -367,23 +369,27 @@ impl AsyncTruncateSinkWriter for PulsarSinkWriter {
         add_future: DeliveryFutureManagerAddFuture<'a, Self::DeliveryFuture>,
     ) -> Result<()> {
         dispatch_sink_formatter_str_key_impl!(&self.formatter, formatter, {
+            // sync: format & own everything
+            let mut pairs: Vec<(Option<String>, Option<Vec<u8>>)> = Vec::new();
```

Comment:
Shall be initialized with `Vec::with_capacity(chunk.cardinality())`.

Besides, what makes a difference in owning it and the later `no formatter/iterator generics across .await`? The formatter variable itself is a reference and should not occupy too much space if it's captured by the future, and furthermore the `Box::pin` should have already broken down the large future.